tests/__init__.py

```diff
@@ -9,69 +10,37 @@
-                "stage": stage,
-                "quality": "high",
-                "kind": kind
-            }
-            if platform:
-                expected_headers["platform"] = platform
-            if extension:
-                expected_headers["extension"] = extension
-
-            payload = {
-                "sample": sample,
-                "extraction_level": 999,
-            }
-            if tag:
-                payload["tags"] = [tag]
-
-            self.assertTasksEqual(results, [Task(expected_headers, payload)])
-
     def test_works_as_expected(self):
+        self.maxDiff = None
         for testcase in os.listdir("tests/testdata"):
+            print(testcase)
+            if not testcase.endswith('.json'):
+                continue
+
+            testcase_config = testcase
+            testcase_content = testcase_config.replace('.json', '')
+
             with self.subTest(testcase):
-                with open(f"tests/testdata/{testcase}", "rb") as f:
+                with open(f"tests/testdata/{testcase_config}", "rb") as f:
+                    expected = json.load(f)
+                with open(f"tests/testdata/{testcase_content}", "rb") as f:
                     content = f.read()
-                expected_tag, file_name = testcase.split("-", 1)
-                tag_elements = expected_tag.split(":")
-
-                # "misc" prefix is added only for mwdb compatibility
-                if tag_elements[0] == "misc":
-                    tag_elements = tag_elements[1:]
 
-                headers = {
-                    "kind": tag_elements[0],
-                    'stage': "recognized",
+                sample = TestResource(testcase_content, content)
+                task = Task({
+                    "type": "sample",
+                    "kind": "raw"
+                }, payload={
+                    "sample": sample,
+                    "extraction_level": 999,
+                })
+
+                payload = {
+                    "sample": sample,
+                    "extraction_level": 999,
                 }
-                if len(tag_elements) > 1:
-                    if tag_elements[0] == "archive":
-                        headers["extension"] = tag_elements[1]
-                    elif tag_elements[0] == "unknown":
-                        expected_tag = None
-                        headers['stage'] = "unrecognized"
-                    else:
-                        headers["platform"] = tag_elements[1]
-                        if len(tag_elements) > 2:
-                            headers["extension"] = tag_elements[2]
-                self._single_file_test(
-                    file_name, content, expected_tag, **headers
-                )
+                if expected.get("payload"):
+                    payload.update(expected["payload"])
+
+                res = self.run_task(task)
+                self.assertTasksEqual(res, [Task(expected["headers"], payload)])
```

Comment:
This looks quite simple and cool, but I'm not sure if hardcoding libmagic output in test cases is a good idea - it changes between versions. These tests are quite broken on my system (file v5.40). Examples:
```
AssertionError: 'PDF document, version 1.4, 1 pages' != 'PDF document, version 1.4'
- PDF document, version 1.4, 1 pages
?                          ---------
+ PDF document, version 1.4
 : Incorrect value of payload.magic
 ```
 ```
 AssertionError: 'ASCII text, with very long lines (32088), with no line terminators' != 'ASCII text, with very long lines, with no line terminators'
- ASCII text, with very long lines (32088), with no line terminators
?                                 --------
+ ASCII text, with very long lines, with no line terminators
```
```
AssertionError: 'Zip archive data, at least v2.0 to extract, compression method=deflate' != 'Zip archive data, at least v2.0 to extract'
- Zip archive data, at least v2.0 to extract, compression method=deflate
?                                           ----------------------------
+ Zip archive data, at least v2.0 to extract
```

Reply:
Yes, that is to be expected. Something like `libmagic` behaves differently, sometimes even within the same version (e.g. with local magic files, etc.). That's why in my experience we choose a frame of reference which is used for the test execution. In this case, this frame of reference is the CI/CD pipeline which runs the tests (based on Ubuntu latest in this case). Then everyone uses that frame of reference for development. This is proven to work and guarantees stable builds.

Reply:
NAK from me. Tests that rely on a specific environment are a bad idea. Apart from annoyances, such as inability to run the tests on a development machine, there are some other problems:

1. Using ubuntu-latest for testing basically guarantees that the environment will be different at some point. When GitHub changes latest from 20.04 to 21.04, tests will probably break unexpectedly and require fixing.
2. Our CI is not and should not be the only testing environment (the more testing we get, the better). At this point we already have a downstream - nixpkgs that perform classifier tests in their CI infrastructure.

Reply:
I can understand your point of view. So what do you suggest?

Reply:
Since the tests require some more consideration I would suggest splitting them off to a separate PR.
We could merge the MIME / magic additions now and take some more time designing the implementation of test suite.

Reply:
"writing the tests later" == "writing the tests never". Maybe we can fix it somehow after all/